go/vt/vttablet/common/config.go

```diff
@@ -0,0 +1,253 @@
+	"time"
+)
+
+type VReplicationConfig struct {
+	ExperimentalFlags       int64
+	NetReadTimeout          int
+	NetWriteTimeout         int
+	CopyPhaseDuration       time.Duration
+	RetryDelay              time.Duration
+	MaxTimeToRetryError     time.Duration
+	RelayLogMaxSize         int
+	RelayLogMaxItems        int
+	ReplicaLagTolerance     time.Duration
+	HeartbeatUpdateInterval int
+	StoreCompressedGTID     bool
+	ParallelInsertWorkers   int
+	TabletTypesStr          string
+
+	VStreamPacketSize                      int
+	VStreamPacketSizeOverride              bool
+	VStreamDynamicPacketSize               bool
+	VStreamDynamicPacketSizeOverride       bool
+	VStreamBinlogRotationThreshold         int64
+	VStreamBinlogRotationThresholdOverride bool
+
+	Overrides map[string]string
+}
+
+var configMutex sync.Mutex
+var DefaultVReplicationConfig *VReplicationConfig
+
+func GetVReplicationConfigDefaults(useCached bool) *VReplicationConfig {
+	configMutex.Lock()
+	defer configMutex.Unlock()
+	if useCached && DefaultVReplicationConfig != nil {
+		return DefaultVReplicationConfig
+	}
+	DefaultVReplicationConfig = &VReplicationConfig{
+		ExperimentalFlags:       vreplicationExperimentalFlags,
+		NetReadTimeout:          vreplicationNetReadTimeout,
+		NetWriteTimeout:         vreplicationNetWriteTimeout,
+		CopyPhaseDuration:       vreplicationCopyPhaseDuration,
+		RetryDelay:              vreplicationRetryDelay,
+		MaxTimeToRetryError:     vreplicationMaxTimeToRetryError,
+		RelayLogMaxSize:         vreplicationRelayLogMaxSize,
+		RelayLogMaxItems:        vreplicationRelayLogMaxItems,
+		ReplicaLagTolerance:     vreplicationReplicaLagTolerance,
+		HeartbeatUpdateInterval: vreplicationHeartbeatUpdateInterval,
+		StoreCompressedGTID:     vreplicationStoreCompressedGTID,
+		ParallelInsertWorkers:   vreplicationParallelInsertWorkers,
+		TabletTypesStr:          vreplicationTabletTypesStr,
+
+		VStreamPacketSizeOverride:              false,
+		VStreamPacketSize:                      VStreamerDefaultPacketSize,
+		VStreamDynamicPacketSizeOverride:       false,
+		VStreamDynamicPacketSize:               VStreamerUseDynamicPacketSize,
+		VStreamBinlogRotationThresholdOverride: false,
+		VStreamBinlogRotationThreshold:         VStreamerBinlogRotationThreshold,
+
+		Overrides: make(map[string]string),
+	}
+	return DefaultVReplicationConfig
+}
+
+func InitVReplicationConfigDefaults() *VReplicationConfig {
+	return GetVReplicationConfigDefaults(true)
+}
+
+func NewVReplicationConfig(overrides map[string]string) (*VReplicationConfig, error) {
+	InitVReplicationConfigDefaults()
+
+	configMutex.Lock()
+	c := &VReplicationConfig{}
+	*c = *DefaultVReplicationConfig
+	configMutex.Unlock()
+
+	c.Overrides = maps.Clone(overrides)
+	var errors []string
+	getError := func(k, v string) string {
+		return fmt.Sprintf("invalid value for %s: %s", k, v)
+	}
+	for k, v := range overrides {
+		if v == "" {
+			continue
+		}
+		switch k {
+		case "vreplication_experimental_flags":
```

Comment:
How will developers know to add a new tablet flag here? I don't think they will, but that's also OK (new flags should be rare). I was just curious.

Reply:
This is an good question. At the moment developers will need to consciously add any new flags here. I started with a plugin based framework to define flags to enforce this but it quickly became overly complex and not necessarily simplifying the code overall.